src/input/input_python.rs

```diff
@@ -352,6 +354,56 @@ impl<'a> Input<'a> for PyAny {
         }
     }
 
+    fn strict_decimal(&'a self, decimal_type: &'a PyType) -> ValResult<&'a PyAny> {
+        let py = decimal_type.py();
+
+        // Fast path for existing decimal objects
+        if self.is_exact_instance(decimal_type) {
+            return Ok(self);
+        }
+
+        // Try subclasses of decimals, they will be upcast to Decimal
+        if self.is_instance(decimal_type)? {
+            match decimal_type.call1((self,)) {
+                Ok(decimal) => return Ok(decimal),
+                Err(e)
+                    if e.matches(
+                        py,
+                        PyTuple::new(
+                            py,
+                            [
+                                py.import("decimal")?.getattr("DecimalException")?,
+                                PyTypeError::type_object(py),
+                            ],
+                        ),
+                    ) => {}
+                Err(e) => return Err(ValError::InternalErr(e)),
+            }
+        }
+
+        Err(ValError::new(
+            ErrorType::IsInstanceOf {
+                class: decimal_type.name().unwrap_or("Decimal").to_string(),
+            },
+            self,
+        ))
+    }
+
+    fn lax_decimal(&'a self, decimal_type: &'a PyType) -> ValResult<&'a PyAny> {
+        // Fast path for existing decimal objects
+        if self.is_exact_instance(decimal_type) {
+            return Ok(self);
+        }
+
+        if self.is_instance_of::<PyString>() || (self.is_instance_of::<PyInt>() && !self.is_instance_of::<PyBool>()) {
+            create_decimal(self, self, decimal_type)
+        } else if self.is_instance_of::<PyFloat>() {
+            create_decimal(self.str()?, self, decimal_type)
+        } else {
+            Err(ValError::new(ErrorType::DecimalType, self))
+        }
+    }
+
```

Comment:
It looks to me like this will error for instances of subclasses of Decimal, not just upcast them.

But I don't see downside to just allowing proper subclass instances through without modification in strict _or_ lax mode. Maybe I'm missing something?

Reply:
Good catch, thanks. I've now fixed the logic to allow subclasses here.

As per https://github.com/pydantic/pydantic-core/pull/763#discussion_r1285650677 I will upcast them for now.